include/exadg/postprocessor/time_control.cpp

```diff
@@ -0,0 +1,142 @@
+    trigger_interval(-1.0),
+    trigger_every_time_steps(dealii::numbers::invalid_unsigned_int)
+{
+}
+
+TimeControl::TimeControl() : EPSILON(1.0e-10), reset_counter(true), counter(0)
+{
+}
+
+// TODO: should not be used anymore!
+void
+TimeControl::reinit(bool const         is_active_,
+                    double const       start_time_,
+                    double const       end_time_,
+                    double const       trigger_interval_,
+                    unsigned int const trigger_every_time_steps_)
+{
+  time_control_data.is_active                = is_active_;
+  time_control_data.start_time               = start_time_;
+  time_control_data.end_time                 = end_time_;
+  time_control_data.trigger_interval         = trigger_interval_;
+  time_control_data.trigger_every_time_steps = trigger_every_time_steps_;
+}
+
+void
+TimeControl::reinit(TimeControlData const & time_control_data_in)
+{
+  time_control_data = time_control_data_in;
+}
+
+// If reinit() is not called needs_evaluation() will always evaluate to false
+bool
+TimeControl::needs_evaluation(double const time, types::time_step const time_step_number) const
+{
+  if(time_control_data.is_active == false)
+    return false;
+
+  // steady case
+  if(!Utilities::is_unsteady_timestep(time_step_number))
+  {
+    ++counter;
+    return true;
+  }
+
+  // unsteady evaluation
+  if(time < time_control_data.start_time || time > time_control_data.end_time)
+    return false;
+
+  bool evaluate = false;
+  if(time_control_data.trigger_every_time_steps > 0)
+  {
+    AssertThrow(time_control_data.trigger_interval < 0.0,
+                dealii::ExcMessage("Input parameters are in conflict."));
+
+    if((time_step_number - 1) % time_control_data.trigger_every_time_steps == 0)
+    {
+      evaluate = true;
+    }
+  }
+  else if(time_control_data.trigger_interval > 0.0)
+  {
+    AssertThrow(time_control_data.trigger_every_time_steps == dealii::numbers::invalid_unsigned_int,
+                dealii::ExcMessage("Input parameters are in conflict."));
+
+    // In case of a restart, time might be significantly larger than start_time in the first time
+    // step of the restarted simulation. Hence, we have to reconstruct the counter at the end of the
+    // previous simulation, in order to obtain the correct behavior at the beginning of the
+    // restarted simulation.
+    if(reset_counter)
+    {
+      counter +=
+        int((time - time_control_data.start_time + EPSILON) / time_control_data.trigger_interval);
+      reset_counter = false;
+    }
+
+    if((time >
+        (time_control_data.start_time + counter * time_control_data.trigger_interval - EPSILON)))
+    {
+      evaluate = true;
+      ++counter;
+    }
+  }
+  else
+  {
+    AssertThrow(false,
+                dealii::ExcMessage("Invalid parameters specified. Use either "
+                                   "trigger_interval > 0.0 or trigger_every_time_steps > 0."));
```

Comment:
we have to think about this error message.

Reply:
should be done